fuzz/fuzz.ml

```diff
@@ -51,48 +61,81 @@ let () =
+        check (srcoff < 0 || srcoff > Bytes.length src ||
+               dstoff < 0 || dstoff > Bytes.length src ||
+               len < 0 ||
+               len > Bytes.length src - srcoff ||
+               len > Cstruct.len dst - dstoff)
     );
   add_test ~name:"blit_to_bytes" [cstruct; int; bytes; int; int] (fun src srcoff dst dstoff len ->
       match Cstruct.blit_to_bytes src srcoff dst dstoff len with
-      | () -> ()
-      | exception Invalid_argument _ -> ()
-    );
-  add_test ~name:"memset" [cstruct; int] (fun c x ->
-      Cstruct.memset c x
-    );
-  add_test ~name:"set_len" [cstruct; int] (fun c x ->
-      match Cstruct.set_len c x with
-      | c2 -> check (Cstruct.len c2 <= Cstruct.len c)
+      | () -> check (Cstruct.equal (Cstruct.sub src srcoff len)
+                                   (Cstruct.sub (Cstruct.of_bytes dst) dstoff len))
+      | exception Invalid_argument _ ->
+        check (srcoff < 0 || srcoff > Cstruct.len src ||
+               dstoff < 0 || dstoff > Cstruct.len src ||
+               len < 0 ||
+               len > Cstruct.len src - srcoff ||
+               len > Bytes.length dst - dstoff)
+    );
+  add_test ~name:"memset" [cstruct; int; int] (fun c x i ->
+      guard (i >= 0 && i < Cstruct.len c);
+      Cstruct.memset c x;
+      check (Cstruct.get_uint8 c i = x land 0xff)
+    );
+  add_test ~name:"set_len" [cstruct; int] (fun base len ->
+      match Cstruct.set_len base len with
+      | x ->
+        (* check_within ~base x; *)   (* Maybe deprecate set_len extending the allocation? *)
+        check (Cstruct.len x >= 0);
+        check (Cstruct.len x <= Bigarray.Array1.dim (base.Cstruct.buffer));
+        check (Cstruct.len x = len)
       | exception Invalid_argument _ -> ()
     );
-  add_test ~name:"add_len" [cstruct; int] (fun c x ->
-      match Cstruct.add_len c x with
-      | c2 -> check (Cstruct.len c2 <= Cstruct.len c)
+  add_test ~name:"add_len" [cstruct; int] (fun base len ->
+      match Cstruct.add_len base len with
+      | x ->
+        check (Cstruct.len x >= 0);
+        check (Cstruct.len x <= Bigarray.Array1.dim (base.Cstruct.buffer));
+        check (Cstruct.len x = Cstruct.len base + len)
       | exception Invalid_argument _ -> ()
     );
-  add_test ~name:"split" [cstruct; option int; int] (fun c start len ->
-      match Cstruct.split ?start c len  with
-      | c1, c2 -> check (Cstruct.len c2 <= Cstruct.len c && Cstruct.len c1 <= Cstruct.len c)
+  add_test ~name:"split" [cstruct; option int; int] (fun base start len ->
+      match Cstruct.split ?start base len  with
+      | c1, c2 ->
+        check_within ~base c1;
+        check_within ~base c2;
+        let start = match start with None -> 0 | Some x -> x in
+        check (Cstruct.equal (Cstruct.sub base start len) c1);
+        check (Cstruct.equal (Cstruct.shift base (start + len)) c2)
       | exception Invalid_argument _ -> ()
     );
   add_test ~name:"BE.set_uint64" [cstruct; int] (fun c off ->
```

Comment:
What do you think about adding similar cases for other BE setters and getters (where there are similar bound checks)?

Reply:
I think anyone who has an idea for extra fuzz tests should feel free to make a PR adding them.